File: api/app/application/services/app_config_service.py
from app.interface.schemas.base import ListA2AServerResponse
from app.interface.schemas.base import ListA2AServerItem
from app.domain.models import app_config
from app.domain.models.app_config import App_Config
from app.domain.models.app_config import A2A_ServerConfig
from app.domain.services.tools.mcp import McpClientManager
from app.interface.schemas.base import ListMCPServerItem
from typing import List
from app.domain.models.app_config import Mcp_Config
from app.domain.models.app_config import Agent_Config
from app.domain.models.app_config import LLM_Config
from app.domain.repositories.app_config_repository import AppConfigRepository
from app.domain.services.tools.a2a import A2AClientManager


class AppConfigService:
    def __init__(self, app_config_repository: AppConfigRepository):
        self.app_config_repository = app_config_repository

    def get_app_config(self) -> App_Config:
        return self.app_config_repository.load()

    def get_llm_config(self) -> LLM_Config:
        return self.get_app_config().llm_config

    def get_agent_config(self) -> Agent_Config:
        return self.get_app_config().agent_config

    async def get_mcp_servers(self) -> List[ListMCPServerItem]:
        mcp_config = self.get_app_config().mcp_config
        mcp_servers = []
        mcp_client_manager = McpClientManager(mcp_config)

        try:
            await mcp_client_manager.initialize()
            tools = mcp_client_manager.tools
            for server_name, server_config in mcp_config.mcpServers.items():
                mcp_servers.append(
                    ListMCPServerItem(
                        server_name=server_name,
                        enabled=server_config.enabled,
                        transport=server_config.transport,
                        tools=[tool.name for tool in tools.get(server_name, [])],
                    )
                )
        finally:
            await mcp_client_manager.cleanup()
        return {mcp_servers}

    def update_llm_config(self, llm_config: LLM_Config):
        app_config = self.get_app_config()
        app_config.llm_config = llm_config
        self.app_config_repository.save(app_config)
        return llm_config

    def update_agent_config(self, agent_config: Agent_Config):
        app_config = self.get_app_config()
        app_config.agent_config = agent_config
        self.app_config_repository.save(app_config)
        return agent_config

    def update_and_create_mcp_server(self, mcp_config: Mcp_Config):
        app_config = self.get_app_config()
        app_config.mcp_config.mcpServers.update(mcp_config.mcpServers)
        self.app_config_repository.save(app_config)
        return app_config.mcp_config

    def delete_mcp_server(self, server_name: str):
        app_config = self.get_app_config()

        if server_name not in app_config.mcp_config.mcpServers:
            raise ValueError(f"MCP服务 {server_name} 不存在")

        del app_config.mcp_config.mcpServers[server_name]
        self.app_config_repository.save(app_config)
        return app_config.mcp_config

    def enable_mcp_server(self, server_name: str, enable: bool):
        app_config = self.get_app_config()

        if server_name not in app_config.mcp_config.mcpServers:
            raise ValueError(f"MCP服务 {server_name} 不存在")

        app_config.mcp_config.mcpServers[server_name].enabled = enable
        self.app_config_repository.save(app_config)
        return app_config.mcp_config

    async def get_a2a_servers(self) -> ListA2AServerResponse:
        a2a_config = self.get_app_config().a2a_config
        a2a_servers = []
        a2a_client_manager = A2AClientManager(a2a_config)
        try:
            await a2a_client_manager.initialize()
            for id, agent_card in a2a_client_manager.agent_cards.items():
                a2a_servers.append(
                    ListA2AServerItem(
                        id=id,
                        name=agent_card.get("name"),
                        description=agent_card.get("description"),
                        input_modes=agent_card.get("defaultInputModes"),
                        output_modes=agent_card.get("defaultOutputModes"),
                        streamable=agent_card.get("capabilities", {}).get(
                            "streaming", False
                        ),
                        push_notifications=agent_card.get("capabilities", {}).get(
                            "push_notifications", False
                        ),
                        enabled=agent_card.get("enabled", False),
                    )
                )
        finally:
            await a2a_client_manager.cleanup()
        return {"a2a_servers": a2a_servers}

    def create_a2a_server(self, base_url: str):
        app_config = self.get_app_config()
        app_config.a2a_config.a2a_servers.append(A2A_ServerConfig(base_url=base_url))
        self.app_config_repository.save(app_config)
        return app_config.a2a_config

    def delete_a2a_server(self, id: str):
        app_config = self.get_app_config()
        a2a_servers = app_config.a2a_config.a2a_servers

        # 查找并删除
        for i, server in enumerate(a2a_servers):
            if server.id == id:
                a2a_servers.pop(i)
                self.app_config_repository.save(app_config)
                return app_config.a2a_config

        raise ValueError(f"A2A服务 {id} 不存在")

    def enable_a2a_server(self, id: str, enable: bool):
        app_config = self.get_app_config()
        a2a_servers = app_config.a2a_config.a2a_servers

        # 查找目标服务器
        server = next((s for s in a2a_servers if s.id == id), None)

        if not server:
            raise ValueError(f"A2A服务 {id} 不存在")

        server.enabled = enable
        self.app_config_repository.save(app_config)
        return app_config.a2a_config

File: api-node/src/services/app-config.service.ts
import fs from 'fs/promises';
import path from 'path';
import YAML from 'yaml';
import lockfile from 'proper-lockfile';
import { v4 as uuidv4 } from 'uuid';
import { getSettings } from '../config';
import { AppConfig, A2AConfig, A2AServerConfig, LLMConfig, AgentConfig, McpConfig, McpServerConfig } from '../domain/models';

export class FileAppConfigRepository {
  constructor(private readonly filepath = getSettings().appConfigFilepath) {}

  private resolvePath(): string {
    return path.isAbsolute(this.filepath)
      ? this.filepath
      : path.resolve(process.cwd(), this.filepath);
  }

  private lockPath(): string {
    return `${this.resolvePath()}.lock`;
  }

  load(): AppConfig {
    const content = require('fs').readFileSync(this.resolvePath(), 'utf-8');
    const parsed = YAML.parse(content) as AppConfig;
    parsed.a2a_config.a2a_servers = (parsed.a2a_config.a2a_servers ?? []).map((s: A2AServerConfig) => ({
      ...s,
      id: s.id ?? uuidv4(),
      enabled: s.enabled ?? false,
    }));
    return parsed;
  }

  async save(config: AppConfig): Promise<void> {
    const filePath = this.resolvePath();
    await fs.mkdir(path.dirname(filePath), { recursive: true });
    const release = await lockfile.lock(filePath, {
      lockfilePath: this.lockPath(),
      retries: { retries: 5, minTimeout: 100, maxTimeout: 1000 },
    });
    try {
      await fs.writeFile(filePath, YAML.stringify(config), 'utf-8');
    } finally {
      await release();
    }
  }
}

export class AppConfigService {
  constructor(private readonly repo = new FileAppConfigRepository()) {}

  getAppConfig(): AppConfig {
    return this.repo.load();
  }

  getLlmConfig(): LLMConfig {
    return this.getAppConfig().llm_config;
  }

  getAgentConfig(): AgentConfig {
    return this.getAppConfig().agent_config;
  }

  getMcpConfig(): McpConfig {
    return this.getAppConfig().mcp_config;
  }

  getA2aConfig(): A2AConfig {
    return this.getAppConfig().a2a_config;
  }

  async updateLlmConfig(config: LLMConfig): Promise<LLMConfig> {
    const appConfig = this.getAppConfig();
    appConfig.llm_config = config;
    await this.repo.save(appConfig);
    return config;
  }

  async updateAgentConfig(config: AgentConfig): Promise<AgentConfig> {
    const appConfig = this.getAppConfig();
    appConfig.agent_config = config;
    await this.repo.save(appConfig);
    return config;
  }

  async getMcpServers() {
    const config = this.getAppConfig();
    return Object.entries(config.mcp_config.mcpServers).map(([serverName, serverConfig]: [string, McpServerConfig]) => ({
      server_name: serverName,
      enabled: serverConfig.enabled,
      transport: serverConfig.transport,
      tools: [] as string[],
    }));
  }

  async updateAndCreateMcpServer(mcpConfig: McpConfig): Promise<McpConfig> {
    const appConfig = this.getAppConfig();
    appConfig.mcp_config.mcpServers = {
      ...appConfig.mcp_config.mcpServers,
      ...mcpConfig.mcpServers,
    };
    await this.repo.save(appConfig);
    return appConfig.mcp_config;
  }

  async deleteMcpServer(serverName: string): Promise<McpConfig> {
    const appConfig = this.getAppConfig();
    if (!(serverName in appConfig.mcp_config.mcpServers)) {
      throw new Error(`MCP服务 ${serverName} 不存在`);
    }
    delete appConfig.mcp_config.mcpServers[serverName];
    await this.repo.save(appConfig);
    return appConfig.mcp_config;
  }

  async enableMcpServer(serverName: string, enable: boolean): Promise<McpConfig> {
    const appConfig = this.getAppConfig();
    if (!(serverName in appConfig.mcp_config.mcpServers)) {
      throw new Error(`MCP服务 ${serverName} 不存在`);
    }
    appConfig.mcp_config.mcpServers[serverName].enabled = enable;
    await this.repo.save(appConfig);
    return appConfig.mcp_config;
  }

  async getA2aServers() {
    const config = this.getAppConfig();
    return {
      a2a_servers: config.a2a_config.a2a_servers.map((s: A2AServerConfig) => ({
        id: s.id!,
        name: '',
        description: '',
        input_modes: [],
        output_modes: [],
        streamable: false,
        push_notifications: false,
        enabled: s.enabled ?? false,
      })),
    };
  }

  async createA2aServer(baseUrl: string): Promise<A2AConfig> {
    const appConfig = this.getAppConfig();
    appConfig.a2a_config.a2a_servers.push({ id: uuidv4(), base_url: baseUrl, enabled: false });
    await this.repo.save(appConfig);
    return appConfig.a2a_config;
  }

  async deleteA2aServer(id: string): Promise<A2AConfig> {
    const appConfig = this.getAppConfig();
    const idx = appConfig.a2a_config.a2a_servers.findIndex((s: A2AServerConfig) => s.id === id);
    if (idx < 0) throw new Error(`A2A服务 ${id} 不存在`);
    appConfig.a2a_config.a2a_servers.splice(idx, 1);
    await this.repo.save(appConfig);
    return appConfig.a2a_config;
  }

  async enableA2aServer(id: string, enable: boolean): Promise<A2AConfig> {
    const appConfig = this.getAppConfig();
    const server = appConfig.a2a_config.a2a_servers.find((s: A2AServerConfig) => s.id === id);
    if (!server) throw new Error(`A2A服务 ${id} 不存在`);
    server.enabled = enable;
    await this.repo.save(appConfig);
    return appConfig.a2a_config;
  }
}

export const appConfigService = new AppConfigService();
